ros_ws/src/get_user_request.py

```diff
@@ -0,0 +1,141 @@
+import os
+import langchain
+
+from langchain.llms import OpenAI
+from langchain import PromptTemplate
+from langchain import schema
+from langchain.prompts import (ChatPromptTemplate,PromptTemplate,SystemMessagePromptTemplate, HumanMessagePromptTemplate)
+from langchain.chat_models import ChatOpenAI
+from langchain.chains import LLMChain
+from langchain.schema import HumanMessage, SystemMessage, AIMessage
+api_key=os.environ["OPENAI_API_KEY"]
+model_to_use="gpt-4-0613"
+#model_to_use="text-davinci-003"
+
+
+def Refine(request):
+    chat = ChatOpenAI(model_name=model_to_use)
+
+    sys_msg=SystemMessage(content="""You are a home assistant robot that can retrieve objects for your user. 
+        You are looking for a single item to search for.  
+        You should ask clarifying questions of your user, only if needed, until you have a 
+        sufficient description of a specific item to search for and retrieve.
+        If the item seems vague or general or the user isn't sure, make a suggestion.
+        If the user requests "something" or "anything", you shouold make a suggestion that is consistent
+        with the rest of the request.  For example, if the user requests please bring me anything cold,
+        you might suggest "would you like a glass of water?" """)
+   
+    unknown_request=True
+    context=[sys_msg]
+    attempts_to_understand=0
+    user_msg="User: " + request
+
+    print (user_msg)
+
+    #If response sufficiently refined 
+    #break out and pass to publish
+    judgement=Judge(request).lower()
+    if judgement.lower()=="yes":
+        #pubish response
+        target=extract_request(request)
+
+        print("The target is ",target)
+        unknown_request=False
+    while unknown_request:
+        #add user request or claification to the conversation
+        context.append(HumanMessage(content=user_msg))
+        response=chat(context)
+        print( response.content)
+        attempts_to_understand+=1
+
+        #seek input from user 
+        request=input("User:")
+        #If response sufficiently refined 
+        #break out and pass to publish
+        judgement=Judge(request).lower()
+        if request.lower()=="yes":
+            target=extract_request(response.content)
+            print("The target is ",target)
+            break
+        if judgement=="yes":
+            #identify simple target
+            target=extract_request(request)
+            #pubish target
+            print("The target is ",target)
+            break
+        #if there have been too many cycles restart the request
+        if attempts_to_understand==5:
+            print("I'm sorry i dont inderstand.  Could we start over?")
+            break
+        context.append(response)
+
+def Judge(request):
+    #llm = OpenAI(model_name=model_to_use, temperature=0)
```

Comment:
If it's commented out it's extraneous code.

Reply:
fair.  This was another way of opening the model whrn I was using davinci.  With gpt4 its ChatOpenAI instead of just OpenAI.